source/base/discrete_time.cc

```diff
@@ -0,0 +1,78 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2019 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+
+#include <deal.II/base/discrete_time.h>
+
+#include <cmath>
+#include <limits>
+
+DEAL_II_NAMESPACE_OPEN
+
+namespace internal
+{
+  // Helper function that computes the next discrete time, adjusting it if:
+  //  - The next time exceeds the end time.
+  //  - The next time is smaller but very close to the end time.
+  double
+  get_next_time(const double current_time,
+                const double step_size,
+                const double end_time)
+  {
+    Assert(end_time >= current_time, ExcInternalError());
+    double           next_time          = current_time + step_size;
+    constexpr double relative_tolerance = 0.05;
```

Comment:
I think the algorithm here should be documented somewhere in the documentation of the class.

Reply:
I added more explanation on the algorithm.